### DISCUSSION TITLE
UserData for small `Copy` objects

### DISCUSSION THREAD

**Discussion by setzer22** [mlua-rs/mlua#132] [Q&A] [answered]
Hey! :wave: I'm trying to come up with a scripting API for an application, and `mlua` so far looks *really* nice :) 

The thing is, I need a way to expose simple `Copy` types like a `Vec3` (i.e. a struct holding 3 floats) to Lua, ideally with support for operator overload. I was thinking about doing this as user data, implementing the `UserData` trait, adding metamethods and so on. But after some benchmarking, I was a bit bummed to see this approach is several orders of magnitude slower than simply doing the equivalent numerical manipulation with individual float variables (i.e., sending x, y and z separately instead of a `Vec3` user data).

So it looks like the userdata solution incurs too much overhead, but I'm not aware of any other mechanism to do something like this. So I'd like to ask if there's something a bit more lightweight than userdata in `mlua` for exposing this kind of "cheap to copy" objects to the lua side.

**Comment by khvzak** [comment, marked as answer]
You could implement [`ToLua`](https://docs.rs/mlua/0.7.4/mlua/trait.ToLua.html) and  [`FromLua`](https://docs.rs/mlua/0.7.4/mlua/trait.FromLua.html) traits for your type and store a struct holding N floats in a lua table (with overloaded operators via metatable).

**Comment by setzer22** [comment]
After some benchmarking of this solution, it looks like the table creation overhead is comparable to the overhead of creating a new userdata. But once the vectors are in Lua, as tables, the more computations that are made in Lua code, the more the vector math is optimized by runtimes like LuaJit, whereas userdata will take the same performance hit on every vector operation.

So indeed, it looks like tables with metamethods ended up being better :)
